packages/patrol_cli/lib/src/features/drive/test_runner.dart

```diff
@@ -39,24 +65,52 @@ class TestRunner {
     if (_devices.isEmpty) {
       throw StateError('no devices to run tests on');
     }
-    if (_tests.isEmpty) {
-      throw StateError('no tests to run');
+    if (_targets.isEmpty) {
+      throw StateError('no test targets to run');
+    }
+
+    final builder = _builder;
+    if (builder == null) {
+      throw StateError('builder is null');
+    }
+
+    final executor = _executor;
+    if (executor == null) {
+      throw StateError('executor is null');
     }
 
     _running = true;
 
     final testRunsOnAllDevices = <Future<void>>[];
     for (final device in _devices.values) {
       Future<void> runTestsOnDevice() async {
-        for (final test in _tests) {
-          await test(device);
+        for (final target in _targets) {
+          if (_disposed) {
+            continue;
+          }
+
+          await builder(target, device);
+
+          for (var i = 0; i < _repeats; i++) {
+            if (_disposed) {
+              continue;
+            }
+
+            await executor(target, device);
+          }
         }
       }
 
-      testRunsOnAllDevices.add(runTestsOnDevice());
+      final futures = runTestsOnDevice();
```

Comment:
Why "futures" if its a single future? Did you mean to wrap the iterated `_targets` in `Future.wait`?

Reply:
Nope, it's a naming mistake, it should be named simply `future`.